### PR TITLE
Cache "Connection was closed" as a singleton

### PR DESCRIPTION
**Pull request by testn**
Eliminate the stacktrace generation to improve performance in high connection/disconnection rate case. 
Signed-off-by: Testo Nakada test1@doramail.com

### CODE REVIEW

File: src/main/java/io/vertx/core/http/impl/ClientConnection.java [hunk 1]
```diff
@@ -74,6 +76,11 @@
   private HttpClientRequestImpl requestForResponse;
   private WebSocketImpl ws;
 
+  static {
+    CONNECTION_CLOSED_EXCEPTION = new VertxException("Connection was closed");
+    CONNECTION_CLOSED_EXCEPTION.setStackTrace(new StackTraceElement[0]);
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/http/impl/ClientConnection.java]: I'm a bit worried by the absence of stack trace in the exception. Could you give a reason for this ?
(I understand that it is required to build the singleton)

**Comment by cescoffier** [comment]
Any particular reason why you want to have a singleton ? Did you see performance hogs ?

**Comment by gmagniez** [comment]
I've opened a bugzilla issue few days ago for something similar:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=483098

**Comment by vietj** [comment]
we should use a real exception like now but instead check we really need to instantiate it (i.e check requests.size() > 0 etc...)

**Comment by testn** [comment]
I saw the performance hog in a short request with a lot of connection/disconnection from a lot of clients. We don't really need real exception either as "Connection was closed" is only thrown by handleEnded() request so the stacktrace is not very important. @gmagniez may be able to attest to some of that

**Comment by vietj** [comment]
It is now actually the case as it was reported to be an issue with JIT in some cases.

**Pull request closed without merging**